### PR TITLE
Refactor/clarify unique and shared variables

### PR DESCRIPTION
**Pull request by Xiangyu-Hu**
unified implement of variable for particle and mesh methods.

### CODE REVIEW

**Comment by ChiZhangatTUM** [comment]
Looks like very fundamental changes has mad in this PR. I would like to have very careful review.

**Comment by Xiangyu-Hu** [comment]
> Looks like very fundamental changes has mad in this PR. I would like to have very careful review.

@ChiZhangatTUM It is not fundamental change but small generalization of variables.

**Review by ChiZhangatTUM — APPROVED** [review]

File: src/shared/particles/base_particles.hpp [hunk 1]
```diff
@@ -35,400 +35,369 @@
...
-        constexpr int type_index = DataTypeIndex<VariableType>::value;
+        constexpr int type_index = DataTypeIndex<DataType>::value;
+        return std::get<type_index>(all_particle_data_)[variable->IndexInContainer()];
+    }
 
-        if (all_variable_maps_[type_index].find(variable_name) != all_variable_maps_[type_index].end())
-            return std::get<type_index>(all_particle_data_)[all_variable_maps_[type_index][variable_name]];
+    std::cout << "\nErr0r: the variable '" << variable_name << "' is not registered!\n";
+    std::cout << __FILE__ << ':' << __LINE__ << std::endl;
+    exit(1);
```
Comment (FabienPean-Virtonomy) [inline src/shared/particles/base_particles.hpp:137]:
```suggestion
```

File: src/shared/particles/base_particles.hpp [hunk 2]
```diff
@@ -35,400 +35,369 @@
...
-        {
-            if (variable->getName() == variable_name)
-            {
-                isRegistered = true;
-                return variable->getValue();
-            }
-        }
+    std::cout << "\nError: the variable '" << variable_name << "' is not registered!\n";
+    std::cout << __FILE__ << ':' << __LINE__ << std::endl;
+    exit(1);
```
Comment (FabienPean-Virtonomy) [inline src/shared/particles/base_particles.hpp:99]:
```suggestion
```

File: src/shared/bodies/base_body.h [hunk 3]
```diff
@@ -35,191 +35,185 @@
...
-		{
-			base_particles_->template addDerivedVariableToWrite<DerivedVariableMethod>(std::forward<Args>(args)...);
-		};
-
-        template <class DerivedVariableMethod,typename... Args>
-		void addDerivedBodyState(Args&&... args)
-		{
-			base_particles_->template addDerivedVariable<DerivedVariableMethod>(std::forward<Args>(args)...);
-		};
-
```
Comment (FabienPean-Virtonomy) [inline src/shared/bodies/base_body.h:173]: What is the better way to add a derived variable after this PR, directly adding it to the BaseParticles ?

**Review by FabienPean-Virtonomy — CHANGES_REQUESTED** [review]

File: src/shared/particles/base_particles.hpp (hunk 1 above)
Comment (Xiangyu-Hu) [reply src/shared/particles/base_particles.hpp:137]: we need to exit the code if the error happens.

File: src/shared/particles/base_particles.hpp (hunk 2 above)
Comment (Xiangyu-Hu) [reply src/shared/particles/base_particles.hpp:99]: we need to exit the code if the error happens.

File: src/shared/bodies/base_body.h (hunk 3 above)
Comment (Xiangyu-Hu) [reply src/shared/bodies/base_body.h:173]: Yes. As data output is done in base particles.

File: src/shared/particles/base_particles.hpp (hunk 1 above)
Comment (FabienPean-Virtonomy) [reply src/shared/particles/base_particles.hpp:137]: No, it is not needed. You are forcefully crashing the program that someone could easily recover from. It is sufficient to return a nullptr like it was before.

Example use-case:

```cpp
if(auto* ptr = getGlobalVariableByName<Vec3d>("Position"))
{
   // do something with ptr
}
```
I updated the suggestions accordingly to restore previous behavior

File: src/shared/particles/base_particles.hpp (hunk 2 above)
Comment (FabienPean-Virtonomy) [reply src/shared/particles/base_particles.hpp:99]: See updated suggestion and comment above

**Comment by Xiangyu-Hu** [comment]
@FabienPean-Virtonomy could you have a check the revision according to you suggestion and approve the pull request?

**Review by FabienPean-Virtonomy — APPROVED** [review]

**Pull request merged**